src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkBatchInsertRecord.java

```diff
@@ -0,0 +1,329 @@
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.sql.Types;
+import java.text.DecimalFormat;
+import java.text.MessageFormat;
+import java.time.OffsetDateTime;
+import java.time.OffsetTime;
+import java.time.format.DateTimeFormatter;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map.Entry;
+import java.util.Set;
+
+/**
+ * A simple implementation of the ISQLServerBulkRecord interface that can be used to read in the basic Java data types from an ArrayList of Parameters
+ * that were provided by pstmt/cstmt.
+ */
+public class SQLServerBulkBatchInsertRecord extends SQLServerBulkCommon implements ISQLServerBulkRecord, java.lang.AutoCloseable {
+
+    private ArrayList<Parameter[]> batchParam;
+    private int batchParamIndex = -1;
+    private ArrayList<String> columnList;
+    private ArrayList<String> valueList;
+
+    public SQLServerBulkBatchInsertRecord(ArrayList<Parameter[]> batchParam,
+            ArrayList<String> columnList,
+            ArrayList<String> valueList,
+            String encoding) throws SQLServerException {
+        loggerClassName = "com.microsoft.sqlserver.jdbc.SQLServerBulkBatchInsertRecord";
+        loggerExternal = java.util.logging.Logger.getLogger(loggerClassName);
+        loggerExternal.entering(loggerClassName, "SQLServerBulkBatchInsertRecord", new Object[] {batchParam, encoding});
+
+        if (null == batchParam) {
+            throwInvalidArgument("batchParam");
+        }
+
+        if (null == valueList) {
+            throwInvalidArgument("valueList");
+        }
+
+        this.batchParam = batchParam;
+        this.columnList = columnList;
+        this.valueList = valueList;
+        columnMetadata = new HashMap<>();
+
+        loggerExternal.exiting(loggerClassName, "SQLServerBulkBatchInsertRecord");
+    }
+
+    /**
+     * Releases any resources associated with the batch.
+     * 
+     * @throws SQLServerException
+     *             when an error occurs
+     */
+    public void close() throws SQLServerException {
+    }
+
+    public DateTimeFormatter getColumnDateTimeFormatter(int column) {
+        return columnMetadata.get(column).dateTimeFormatter;
+    }
+
+    @Override
+    public Set<Integer> getColumnOrdinals() {
+        return columnMetadata.keySet();
+    }
+
+    @Override
+    public String getColumnName(int column) {
+        return columnMetadata.get(column).columnName;
+    }
+
+    @Override
+    public int getColumnType(int column) {
+        return columnMetadata.get(column).columnType;
+    }
+
+    @Override
+    public int getPrecision(int column) {
+        return columnMetadata.get(column).precision;
+    }
+
+    @Override
+    public int getScale(int column) {
+        return columnMetadata.get(column).scale;
+    }
+
+    @Override
+    public boolean isAutoIncrement(int column) {
+        return false;
+    }
+
+    @Override
+    public Object[] getRowData() throws SQLServerException {
```

Comment:
There are a few things which can be done better here.

1. Currently loops twice for, this is actually unnecessary, we can do everything here in 1 loop.
2. Stores a lot of one time use information in arrays. Very performance unfriendly if our row/column count is huge.
3. Very very hard to read. Between the massive switch, parsing logic, and referencing 2 arrays and a set; it's difficult to figure out what it's doing.

Proposed Changes:

1. Move the massive switch statement to it's own function.
2. Move the parsing to its own function (this one might be difficult with how `valueIndex` is updated)
3. Change the iteration logic
3a. Create an Object[] to return
3b. Iterate over the set with foreach
3c. Create an Object to be inserted into our Object[]; use the Set as index
3d. Parse the string from valueList with this index
3e. Take the string from the above step and convert it to a desired value through the switch statement
3f. Store this value into our Object, and then add our Object into our Object[].

I've done a "prototype" implementation here: [gist link](https://gist.github.com/rene-ye/b9f0424e90aa6ff73f7c4261cf7162a0). Please take a look and adjust as neccessary.

Reply:
Thanks. I think your implementation is good - I slapped my code onto the existing implementation for the getRowData() method from SQLServerBulkCSVFileRecord, and I didn't double check to see if I can refactor some of the code.

By the way, I decided not to merge the two getRowData() methods together (from SQLServerBulkCSVFileRecord and this file), since the CSV one assumes the data is always String, whereas this one doesn't (which could imply the implementation might change in the future). I'm open to suggestions, though.